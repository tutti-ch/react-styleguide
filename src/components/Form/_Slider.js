import React, { Component } from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import WithWrapper from "./_WithWrapper";
import classes from "./Form.scss";


export class Slider extends Component {
  static defaultProps = {
    template: "",
    minDistance: 10,
    prefix: "",
    suffix: "",
    crossThumbs: false,
    values: [],
    mouseThreshold: 50
  };

  static propTypes = {
    /**
     * The minimum range.
     */
    min: PropTypes.number,

    /**
     * The maximum range.
     */
    max: PropTypes.number,

    values: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ),

    /**
     * If provided, the functionality of the array changes. Step, minValue,
     * maxValue, min, max will be ignored.
     */
    range: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        label: PropTypes.string
      })
    ),

    /**
     * The minumum range in pixels between the two thumbs. This will only
     * work when crossThumbs is false. By default it works.
     */
    minDistance: PropTypes.number,

    /**
     * The prefix for the values. Could be a string or an array.
     * If an array is given, the first value is for the minimum value
     * and the second value is for the maximum value.
     */
    prefix: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),

    /**
     * The suffix for the values. Could be a string or an array.
     * If an array is given, the first value is for the minimum value
     * and the second value is for the maximum value.
     */
    suffix: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),

    /**
     * The step number.
     */
    step: PropTypes.number,

    /**
     * Whether to have two sliders or one.
     */
    multiple: PropTypes.bool,

    /**
     * The label that will be displayed.
     */
    label: PropTypes.string,

    /**
     * Whether min and max thumbs can cross and change positions.
     */
    crossThumbs: PropTypes.bool,

    /**
     * The name of the input. In case of multiple inputs, provide an array with two indexes.
     */
    name: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string)
    ]),

    /**
     * Number of decimals in case step is < 0.
     */
    decimals: PropTypes.number,

    /**
     * The placeholder in case value is empty.
     */
    placeholder: PropTypes.string,

    /**
     * Whether the extremes should return value null or not. Defaults false.
     */
    extremes: PropTypes.bool,

    /**
     * Number of pixels that you need to pull in order to reset filters.
     */
    mouseThreshold: PropTypes.number,
  };

  constructor(props) {
    super(props);

    this.handleMouseDown = this.handleMouseDown.bind(this);
    this.handleMouseUp = this.handleMouseUp.bind(this);
    this.handleMouseMove = this.handleMouseMove.bind(this);
    this.calculatePosition = this.calculatePosition.bind(this);
    this.calculateMousePosition = this.calculateMousePosition.bind(this);
    this.calculateClosestValue = this.calculateClosestValue.bind(this);
    this.getFormattedValue = this.getFormattedValue.bind(this);
    this.getRangeIndex = this.getRangeIndex.bind(this);
    this.getMaxRange = this.getMaxRange.bind(this);
    this.getMinRange = this.getMinRange.bind(this);
    this.renderThumb = this.renderThumb.bind(this);
    this.renderDesc = this.renderDesc.bind(this);

    this.state = {
      min: {
        range: this.getMinRange(),
        input: Array.isArray(props.name) ? props.name[0] : props.name,
        value: parseInt(props.values[0])
      },
      max: {
        range: this.getMaxRange(),
        input: Array.isArray(props.name) ? props.name[1] : props.name,
        value: parseInt(props.values[1])
      },
      prefix: Array.isArray(props.prefix)
        ? props.prefix
        : [props.prefix, props.prefix],
      suffix: Array.isArray(props.suffix)
        ? props.suffix
        : [props.suffix, props.suffix],
      dragging: false // The element we are currently dragging.
    };

    this.state.min.position = this.calculatePosition(this.state.min.value) || 0;
    this.state.max.position =
      this.calculatePosition(this.state.max.value) || 100;
  }

  /**
   * Register the root component here to save some operations later.
   */
  componentDidMount() {
    this.root = ReactDOM.findDOMNode(this);
  }

  componentWillReceiveProps({ name }) {
    const state = {};

    if (name !== this.props.name) {
      state.min = {
        ...this.state.min,
        input: Array.isArray(name) ? name[0] : name
      };
      state.max = {
        ...this.state.max,
        input: Array.isArray(name) ? name[1] : name
      };
    }

    if (Object.keys(state).length) {
      this.setState(state);
    }
  }

  /**
   * Given a value, calculate the position in percentage.
   *
   * @param {number} value
   * @return {number|null} Returns null in case value is also empty.
   */
  calculatePosition(value) {
    if (value === null) return null;

    const minRange = this.state.min.range;
    const maxRange = this.state.max.range;

    if (value === minRange) return 0;
    if (value === maxRange) return 100;

    const total = maxRange - minRange;
    const distance = value - minRange;

    return distance * 100 / total;
  }

  /**
   * Calculate the mouse position in percentage. This function supports also
   * touch events.
   *
   * @param {Object} e The mouse event.
   * @return {number}
   */
  calculateMousePosition(e) {
    const rect = this.root.getBoundingClientRect();
    const totalLen = this.root.offsetWidth;
    const clientX = Slider.clientX(e);
    return Math.max(Math.min((clientX - rect.left) * 100 / totalLen, 100), 0);
  }

  /**
   * Return the clientX value of the event.
   *
   * @param e
   * @return {*}
   */
  static clientX(e) {
    if (e.clientX) return e.clientX;
    if (e.touches && e.touches[0]) return e.touches[0].clientX;
  }

  /**
   * Returns true if value is NaN or a null value.
   *
   * @param val
   * @return {boolean}
   */
  static isEmpty(val) {
    return isNaN(val) || val === null || typeof val === "undefined";
  }

  /**
   * Calculate the closest step value to the value of the given prop.
   * For instance if step is 100, a value of 140 will be rounded to 100 and
   * a value of 170 will be rounded to 200 (as long as they are within min and
   * max limits).
   *
   * @param {string|number} value
   * @return {number}
   */
  calculateClosestValue(value) {
    const { step, range = [] } = this.props;

    let prevValue;
    let nextValue;

    if (range.length === 0) {
      const min = this.state.min.range;
      const max = this.state.max.range;
      prevValue = Math.max(value - (value - min) % step, min); // This calc gives us the prev value
      nextValue = Math.min(prevValue + step, max); // prevValue + step is the next value
    } else {
      const sorted = range.map(i => +i.value).sort((a, b) => a - b);
      prevValue = sorted.filter(i => i <= value).pop();
      nextValue = sorted.filter(i => i > value).shift();
    }

    if (this.direction === "R") {
      return value - prevValue >= nextValue - value ? nextValue : prevValue;
    } else {
      return value - prevValue > nextValue - value ? nextValue : prevValue;
    }
  }

  /**
   * Set the element which has triggered the mouse down event as the
   * element being dragged. Also this event will register the global
   * events for dragging.
   *
   * @param e
   */
  handleMouseDown(e) {
    this.setState({ dragging: e.target });
    this.clientX = Slider.clientX(e);

    e.target.classList.remove(classes.trans);

    // istanbul ignore else
    if (window && window.addEventListener) {
      window.addEventListener("mouseup", this.handleMouseUp);
      window.addEventListener("touchend", this.handleMouseUp);
      window.addEventListener("mousemove", this.handleMouseMove);
    }
  }

  /**
   * Calculate the closest step value when the mouse move or touch has ended.
   */
  handleMouseUp() {
    // Remove global listeners
    // istanbul ignore else
    if (window && window.removeEventListener) {
      window.removeEventListener("mouseup", this.handleMouseUp);
      window.removeEventListener("touchend", this.handleMouseUp);
      window.removeEventListener("mousemove", this.handleMouseMove);
    }

    const { dragging: elem } = this.state;
    const prop = elem.getAttribute("name");
    elem.classList.add(classes.trans);

    const value = this.calculateClosestValue(this.state[prop].value);
    const state = this.state[prop];

    this.setState(
      {
        dragging: false,
        [prop]: { ...state, value, position: this.calculatePosition(value) }
      },
      () => {
        if (typeof this.props.onChange === "function") {
          this.props.onChange(this.state[prop].value, {
            name: this.state[prop].input,
            initialValue: this.props.values[prop === "min" ? 0 : 1]
          });
        }
      }
    );
  }

  /**
   * Calculate the cursor position and set the value for the element
   * being dragged.
   *
   * @param {Object} e The event object.
   */
  handleMouseMove(e) {
    // TouchMove throws a warning for e.preventDefault
    if (e.clientX) e.preventDefault();

    // Limits of the slider (minimum and maximum value) and the thumb that is being dragged
    const {
      dragging: elem,
      min: { range: minRange },
      max: { range: maxRange }
    } = this.state;
    const { minDistance, crossThumbs, values, extremes, mouseThreshold } = this.props;

    const prop = elem.getAttribute("name");
    const rect = elem.getBoundingClientRect();
    const isMin = prop === "min";
    const clientX = Slider.clientX(e);
    const mousePos = this.calculateMousePosition(e); // The mouse position in percentage
    const mouseValue = Math.round(
      minRange + (maxRange - minRange) * mousePos / 100
    ); // The value at the mouse position

    // If the mouse position is in the left or right extreme, reset the values (only if initials values are null)
    if (
      isMin &&
      (values[0] === null || extremes) &&
      clientX < rect.left - mouseThreshold
    ) {
      return this.setState({
        min: { ...this.state.min, value: null, position: 0 }
      });
    }

    if (
      !isMin &&
      (values[1] === null || extremes) &&
      clientX > rect.right + mouseThreshold
    ) {
      return this.setState({
        max: { ...this.state.max, value: null, position: 100 }
      });
    }

    // Save the direction so that the calculatePosition
    // function can compute properly
    this.direction = this.clientX < Slider.clientX(e) ? "R" : "L";

    // If cross thumbs is not allowed, make sure that min never passes max thumb.
    if (crossThumbs !== true) {
      const minValue = isMin ? mouseValue : this.state.min.value;
      const maxValue = !isMin ? mouseValue : this.state.max.value;

      if (minValue + minDistance > maxValue) {
        return;
      }
    }

    if (mouseValue >= minRange && mouseValue <= maxRange) {
      const state = {
        ...this.state[prop],
        value: mouseValue,
        position: mousePos
      };

      this.setState({ [prop]: state });
    }
  }

  /**
   * Return the index of the given value in the range.
   *
   * @param value
   * @return {number}
   */
  getRangeIndex(value) {
    const { range = [] } = this.props;

    for (let i = 0; i < range.length; i++) {
      if (+range[i].value === +value) {
        return i;
      }
    }

    return -1;
  }

  /**
   * Return the formatted value.
   *
   * @param val
   * @return {*}
   */
  getFormattedValue(val) {
    if (Slider.isEmpty(val)) {
      return;
    }

    const { range = [], step, decimals = 2 } = this.props;

    if (range.length) {
      return (range.filter(i => +i.value === +val)[0] || {}).label;
    }

    if (step < 1) {
      return val.toFixed(decimals);
    }

    return Math.floor(val);
  }

  /**
   * Return the first non null minimum value that is provided.
   *
   * @return {Number}
   */
  getMinRange() {
    const { range = [] } = this.props;

    // Return first non null value
    if (range[0]) {
      return +range[0].value;
    }

    return +this.props.min;
  }

  /**
   * Return the first non null maximum value that is provided.
   *
   * @return {Number}
   */
  getMaxRange() {
    const { range = [] } = this.props;

    // Return first non null value
    if (range[range.length - 1]) {
      return +range[range.length - 1].value;
    }

    return +this.props.max;
  }

  /**
   * Function to render the thumbs.
   *
   * @param {string} prop minValue|maxValue
   * @return {Array}
   */
  renderThumb(prop) {
    const events = {
      onMouseDown: this.handleMouseDown,
      onTouchStart: this.handleMouseDown,
      onTouchMove: this.handleMouseMove
    };

    const inputName = this.state[prop].input;
    const position = this.state[prop].position;
    const styles = { left: `${position}%` };

    return [
      <span
        className={classes.thumb}
        name={prop}
        style={styles}
        tabIndex={1}
        key={prop}
        {...events}
      />,
      <input
        type="hidden"
        key={`hidden-${prop}`}
        value={this.state[prop]}
        name={inputName}
      />
    ];
  }

  /**
   * Render the description part.
   */
  renderDesc() {
    const { multiple, crossThumbs } = this.props;
    let { min: { value: minValue }, max: { value: maxValue } } = this.state;
    let { prefix, suffix } = this.props;
    let minValueText, maxValueText;

    if (Slider.isEmpty(minValue) && Slider.isEmpty(maxValue)) {
      return this.props.placeholder;
    }

    // Formatter function
    const f = this.getFormattedValue;

    if (multiple) {
      // If crossThumbs is enabled, check for the min value (min and max can be swapped)
      // Otherwise, minValue is always minValue. Do not need to swap here.
      const min = !crossThumbs
        ? minValue
        : minValue < maxValue ? minValue : maxValue;
      const max = !crossThumbs
        ? maxValue
        : maxValue > minValue ? maxValue : minValue;

      // Join prefix and suffixes only when the value is not empty.
      minValueText = min
        ? [prefix[0], f(min), suffix[0]].filter(i => i).join(" ")
        : null;
      maxValueText = max
        ? [prefix[1], f(max), suffix[1]].filter(i => i).join(" ")
        : null;
    } else {
      minValueText = [prefix[0], f(minValue), suffix[0]]
        .filter(i => i)
        .join(" ");
    }

    return [
      minValueText && (
        <span key="minval" className={classes.rangeMin}>
          {minValueText}
        </span>
      ),
      multiple ? (
        <span key="maxval" className={classes.rangeMax}>
          {maxValueText}
        </span>
      ) : null
    ];
  }

  render() {
    const { multiple, label } = this.props;

    return (
      <span className={classes.slider}>
        <span className={classes.label}>
          <span className={classes.labelText}>{label}</span>
          {this.renderDesc()}
        </span>
        {this.renderThumb("min")}
        {multiple && this.renderThumb("max")}
      </span>
    );
  }
}

export default WithWrapper(Slider);
