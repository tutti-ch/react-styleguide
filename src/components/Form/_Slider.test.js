/* global describe, test, expect, jest */
import React from "react"
import { mount } from "enzyme"
import { Slider } from "./_Slider"

describe("(Component) Slider", () => {
  test("[componentDidMount] should register a root element", () => {
    const comp = mount(<Slider min={1000} max={2000}/>)
    const inst = comp.instance()
    expect(inst.root).not.toBeUndefined()
  })

  test("[calculatePosition] should calculate the position given a value", () => {
    const comp = mount(<Slider min={1000} max={2000}/>)
    const inst = comp.instance()
    expect(inst.calculatePosition(1500)).toBe(50)
    expect(inst.calculatePosition(1000)).toBe(0)
    expect(inst.calculatePosition(2000)).toBe(100)
    expect(inst.calculatePosition(1750)).toBe(75)
  })

  test("[calculateMousePosition] should give you the percentage of the mouse position", () => {
    const comp = mount(<Slider min={500} max={1500}/>)
    const inst = comp.instance()
    inst.root = {
      getBoundingClientRect: jest.fn().mockReturnValue({
        left: 500,
      }),
      offsetWidth: 1000,
    }

    expect(inst.calculateMousePosition({ clientX: 750 })).toBe(25)
    expect(inst.calculateMousePosition({ touches: [{ clientX: 1000 }] })).toBe(50)
  })

  test("[calculateClosestValue] should return the closest step value", () => {
    const comp = mount(<Slider min={500} max={1500} step={250}/>)
    const inst = comp.instance()
    comp.setState({ minValue: 800 })
    expect(inst.calculateClosestValue("minValue")).toBe(750)
    comp.setState({ minValue: 650 })
    expect(inst.calculateClosestValue("minValue")).toBe(750)
    comp.setState({ minValue: 1450 })
    expect(inst.calculateClosestValue("minValue")).toBe(1500)
    comp.setState({ minValue: 100000 }) // This should never happen but still let's try
    expect(inst.calculateClosestValue("minValue")).toBe(1500)
    comp.setState({ minValue: 100 }) // This should never happen but still let's try
    expect(inst.calculateClosestValue("minValue")).toBe(500)
  })

  test("[handleMouseDown] should set the element that triggered the event and register variables", () => {
    const comp = mount(<Slider min={500} max={1500} step={250}/>)
    const inst = comp.instance()
    expect(comp.state("dragging")).toBe(false)

    const event = { target: { classList: { remove: jest.fn() } } }
    const spy = jest.spyOn(window, "addEventListener")

    inst.handleMouseDown(event)
    expect(event.target.classList.remove).toHaveBeenCalled()
    expect(comp.state("dragging")).toBe(event.target)
    expect(window.addEventListener).toHaveBeenCalledWith("mouseup", inst.handleMouseUp)
    expect(window.addEventListener).toHaveBeenCalledWith("touchend", inst.handleMouseUp)
    expect(window.addEventListener).toHaveBeenCalledWith("mousemove", inst.handleMouseMove)
    spy.mockRestore()
  })

  test("[handleMouseUp] should remove the listeners and set the position of the element to the nearest step value", () => {
    const comp = mount(<Slider min={500} max={1500} step={250}/>)
    const inst = comp.instance()
    const elem = {
      getAttribute: jest.fn().mockReturnValue("minValue"),
      classList: { add: jest.fn() },
    }

    comp.setState({ dragging: elem, minValue: 770 })
    const spy = jest.spyOn(window, "removeEventListener")

    inst.handleMouseUp()
    expect(elem.classList.add).toHaveBeenCalled()
    expect(comp.state("dragging")).toBe(false)
    expect(comp.state("minValue")).toBe(750)
    expect(window.removeEventListener).toHaveBeenCalledWith("mouseup", inst.handleMouseUp)
    expect(window.removeEventListener).toHaveBeenCalledWith("touchend", inst.handleMouseUp)
    expect(window.removeEventListener).toHaveBeenCalledWith("mousemove", inst.handleMouseMove)
    spy.mockRestore()
  })

  test("[handleMouseMove] should move the element to the mouse position", () => {
    const comp = mount(<Slider min={500} max={1500} step={250} minRange={100}/>)
    const inst = comp.instance()
    const elem = { getAttribute: jest.fn().mockReturnValue("minValue") }
    const event = { preventDefault: jest.fn(), clientX: 570 }
    inst.root = { getBoundingClientRect: jest.fn().mockReturnValue({ left: 470 }), offsetWidth: 1000 }
    comp.setState({ dragging: elem, minValue: 550, maxValue: 750 })

    inst.handleMouseMove(event)
    expect(comp.state("minValue")).toBe(500)

    // Should not let continue when minValue > maxValue - range
    comp.setState({ minValue: 550, maxValue: 650 })
    inst.calculateMousePosition = jest.fn().mockReturnValue(75)
    inst.handleMouseMove(event)
    expect(comp.state("minValue")).toBe(550) // Should remain the same

    elem.getAttribute = jest.fn().mockReturnValue("maxValue")
    inst.handleMouseMove(event)
    expect(comp.state("minValue")).toBe(550)
    expect(comp.state("maxValue")).toBe(1250)

    // The preventDefault should be called every time until now
    expect(event.preventDefault).toHaveBeenCalledTimes(3)
    // Now lets set clientX to undefined to simulate a touch event
    delete event.clientX

    // In this case minValue can be greater than maxValue
    comp.setProps({ crossThumbs: true })
    comp.setState({ minValue: 550, maxValue: 650 })
    elem.getAttribute = jest.fn().mockReturnValue("minValue")
    inst.calculateMousePosition = jest.fn().mockReturnValue(75)
    inst.handleMouseMove(event)
    expect(comp.state("minValue")).toBe(1250)
    expect(event.preventDefault).toHaveBeenCalledTimes(3)

    // Here the mouse value should be greater than max, so we won't call the setState fn
    const stateSpy = jest.spyOn(inst, "setState")
    inst.calculateMousePosition = jest.fn().mockReturnValue(150)
    inst.handleMouseMove(event)
    expect(stateSpy).not.toHaveBeenCalled()
  })

  test("[renderThumb] should choose the correct name for inputs", () => {
    const comp = mount(<Slider min={500} max={1500} step={250} minRange={100} name={["pe", "ps"]} multiple/>)
    const inst = comp.instance()

    expect(mount(<div>{inst.renderThumb("maxValue")}</div>).find("input").prop("name")).toBe("ps")
    expect(mount(<div>{inst.renderThumb("minValue")}</div>).find("input").prop("name")).toBe("pe")

    comp.setProps({ name: "my-name" })
    expect(mount(<div>{inst.renderThumb("minValue")}</div>).find("input").prop("name")).toBe("my-name")
    expect(mount(<div>{inst.renderThumb("maxValue")}</div>).find("input").prop("name")).toBe("my-name")
  })

  test("[handleOnChange] should inform the parent when there is a change on the input", () => {
    const onChange = jest.fn()
    const comp = mount(<Slider min={500} max={1500} step={250} minValue={1000} name={["pe", "ps"]} onChange={onChange}
                               multiple/>)
    const inst = comp.instance()

    inst.handleOnChange("pe", 1000)({ target: { value: 750 } })
    expect(onChange).toHaveBeenCalledWith(750, { name: "pe", initialValue: 1000 })
  })

  describe("snapshots", () => {
    test("should create the right snapshot for single thumb", () => {
      const comp = mount(<Slider min={500} max={1500} step={250} minRange={100}/>)
      expect(comp).toMatchSnapshot()
    })

    test("should create the right snapshot for multiple thumbs", () => {
      const comp = mount(<Slider min={500} max={1500} step={250} minRange={100} multiple/>)
      expect(comp).toMatchSnapshot()
    })

    test("should create the right snapshot for decimals and steps specified", () => {
      const comp = mount(<Slider min={1} max={10} step={0.5} minRange={1} decimals={2} multiple/>)
      expect(comp).toMatchSnapshot()
    })
  })
})
