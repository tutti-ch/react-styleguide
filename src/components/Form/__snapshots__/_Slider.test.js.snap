// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Slider snapshots should create the right snapshot for decimals and steps specified 1`] = `
<Slider
  max={10}
  min={1}
  minDistance={10}
  mouseThreshold={50}
  multiple={true}
  name={
    Array [
      "c",
      "d",
    ]
  }
  prefix=""
  step={0.5}
  suffix=""
  template=""
  values={
    Array [
      2,
    ]
  }
>
  <span>
    <span
      style={
        Object {
          "left": "0%",
          "width": "100%",
        }
      }
    />
    <span>
      <span />
      <span
        key="minval"
      >
        2.00
      </span>
      <span
        key="maxval"
      />
    </span>
    <span>
      <span
        data-name="min"
        key="min"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
        tabIndex={1}
      />
      <input
        key="hidden-min"
        name="c"
        type="hidden"
        value={
          Object {
            "input": "c",
            "range": 1,
            "value": 2,
          }
        }
      />
      <span
        data-name="max"
        key="max"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
          }
        }
        tabIndex={1}
      />
      <input
        key="hidden-max"
        name="d"
        type="hidden"
        value={
          Object {
            "input": "d",
            "range": 10,
            "value": NaN,
          }
        }
      />
    </span>
  </span>
</Slider>
`;

exports[`(Component) Slider snapshots should create the right snapshot for decimals and steps specified 2`] = `
<Slider
  max={10}
  min={1}
  minDistance={10}
  mouseThreshold={50}
  multiple={true}
  name={
    Array [
      "pe",
      "ps",
    ]
  }
  prefix=""
  step={0.5}
  suffix=""
  template=""
  values={
    Array [
      2,
    ]
  }
>
  <span>
    <span
      style={
        Object {
          "left": "0%",
          "width": "100%",
        }
      }
    />
    <span>
      <span />
      <span
        key="minval"
      >
        2.00
      </span>
      <span
        key="maxval"
      />
    </span>
    <span>
      <span
        data-name="min"
        key="min"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
        tabIndex={1}
      />
      <input
        key="hidden-min"
        name="pe"
        type="hidden"
        value={
          Object {
            "input": "pe",
            "range": 1,
            "value": 2,
          }
        }
      />
      <span
        data-name="max"
        key="max"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
          }
        }
        tabIndex={1}
      />
      <input
        key="hidden-max"
        name="ps"
        type="hidden"
        value={
          Object {
            "input": "ps",
            "range": 10,
            "value": NaN,
          }
        }
      />
    </span>
  </span>
</Slider>
`;

exports[`(Component) Slider snapshots should create the right snapshot for multiple thumbs 1`] = `
<Slider
  max={1500}
  min={500}
  minDistance={10}
  mouseThreshold={50}
  multiple={true}
  name={
    Array [
      "a",
      "b",
    ]
  }
  prefix=""
  step={250}
  suffix=""
  template=""
  values={
    Array [
      700,
    ]
  }
>
  <span>
    <span
      style={
        Object {
          "left": "0%",
          "width": "100%",
        }
      }
    />
    <span>
      <span />
      <span
        key="minval"
      >
        700
      </span>
      <span
        key="maxval"
      />
    </span>
    <span>
      <span
        data-name="min"
        key="min"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
        tabIndex={1}
      />
      <input
        key="hidden-min"
        name="a"
        type="hidden"
        value={
          Object {
            "input": "a",
            "range": 500,
            "value": 700,
          }
        }
      />
      <span
        data-name="max"
        key="max"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
          }
        }
        tabIndex={1}
      />
      <input
        key="hidden-max"
        name="b"
        type="hidden"
        value={
          Object {
            "input": "b",
            "range": 1500,
            "value": NaN,
          }
        }
      />
    </span>
  </span>
</Slider>
`;

exports[`(Component) Slider snapshots should create the right snapshot for single thumb 1`] = `
<Slider
  max={1500}
  min={500}
  minDistance={10}
  mouseThreshold={50}
  prefix=""
  step={250}
  suffix=""
  template=""
  values={
    Array [
      700,
    ]
  }
>
  <span>
    <span
      style={
        Object {
          "left": 0,
          "marginLeft": 0,
          "width": "0%",
        }
      }
    />
    <span>
      <span />
      <span
        key="minval"
      >
        700
      </span>
    </span>
    <span>
      <span
        data-name="min"
        key="min"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "0%",
          }
        }
        tabIndex={1}
      />
      <input
        key="hidden-min"
        type="hidden"
        value={
          Object {
            "input": undefined,
            "range": 500,
            "value": 700,
          }
        }
      />
    </span>
  </span>
</Slider>
`;
